File: benchmarks/baseline_bn_loops.f90
module bnmod
  implicit none
contains
  subroutine process_perm(perm, n, hist, nxt, visited)
    integer, intent(in) :: perm(:), n
    integer(8), intent(inout) :: hist(:)
    integer, intent(inout) :: nxt(:), visited(:)
    integer :: twoN, limit, mask, m, i, bit, v, r, cycles, s, cur
    twoN = 2*n
    limit = 1
    do i = 1, n
       limit = limit * 2
    end do
    do mask = 0, limit-1
       m = mask
       do i = 1, n
          bit = m - (m/2)*2
          v = perm(i)
          if (bit == 0) then
             nxt(i) = v
             nxt(n+i) = n+v
          else
             nxt(i) = n+v
             nxt(n+i) = v
          end if
          m = m/2
       end do
       do r = 1, twoN
          visited(r) = 0
       end do
       cycles = 0
       do s = 1, twoN
          if (visited(s) == 0) then
             cycles = cycles + 1
             cur = s
             do while (visited(cur) == 0)
                visited(cur) = 1
                cur = nxt(cur)
             end do
          end if
       end do
       hist(cycles) = hist(cycles) + 1
    end do
  end subroutine process_perm

  recursive subroutine heap(k, perm, n, hist, nxt, visited)
    integer, intent(in) :: k, n
    integer, intent(inout) :: perm(:), nxt(:), visited(:)
    integer(8), intent(inout) :: hist(:)
    integer :: i, t
    if (k == 1) then
       call process_perm(perm, n, hist, nxt, visited)
       return
    end if
    do i = 1, k-1
       call heap(k-1, perm, n, hist, nxt, visited)
       if (mod(k,2) == 0) then
          t = perm(i); perm(i) = perm(k); perm(k) = t
       else
          t = perm(1); perm(1) = perm(k); perm(k) = t
       end if
    end do
    call heap(k-1, perm, n, hist, nxt, visited)
  end subroutine heap
end module bnmod

program bn
  use bnmod
  implicit none
  integer :: n, i, twoN, nargs
  integer, allocatable :: perm(:), nxt(:), visited(:)
  integer(8), allocatable :: hist(:)
  integer(8) :: total
  character(len=32) :: arg
  n = 10
  nargs = command_argument_count()
  if (nargs >= 1) then
     call get_command_argument(1, arg)
     read(arg, *) n
  end if
  twoN = 2*n
  allocate(perm(n), nxt(twoN), visited(twoN), hist(twoN))
  do i = 1, n
     perm(i) = i
  end do
  hist = 0
  call heap(n, perm, n, hist, nxt, visited)
  total = 0
  do i = 1, twoN
     total = total + hist(i)
  end do
  print '(A,I0)', 'total=', total
end program bn

File: benchmarks/baseline_bn_loops.c
#include <stdio.h>
#include <stdlib.h>
static long long hist[41];
static int perm[20], nxt[40], visited[40];
static int N, twoN;
static void process_perm(void){
    int limit=1,b; for(b=0;b<N;b++) limit*=2;
    int mask;
    for(mask=0; mask<limit; mask++){
        int m=mask,i;
        for(i=0;i<N;i++){
            int bit=m-(m/2)*2, v=perm[i];
            if(bit==0){ nxt[i]=v-1; nxt[N+i]=N+v-1; }
            if(bit==1){ nxt[i]=N+v-1; nxt[N+i]=v-1; }
            m/=2;
        }
        int r; for(r=0;r<twoN;r++) visited[r]=0;
        int cycles=0,s;
        for(s=0;s<twoN;s++){
            if(!visited[s]){ cycles++; int cur=s; while(!visited[cur]){visited[cur]=1;cur=nxt[cur];} }
        }
        hist[cycles]++;
    }
}
static void heap(int k){
    if(k==1){ process_perm(); return; }
    int i;
    for(i=0;i<k-1;i++){
        heap(k-1);
        if((k-(k/2)*2)==0){int t=perm[i];perm[i]=perm[k-1];perm[k-1]=t;}
        if((k-(k/2)*2)==1){int t=perm[0];perm[0]=perm[k-1];perm[k-1]=t;}
    }
    heap(k-1);
}
int main(int argc,char**argv){
    N = argc>1?atoi(argv[1]):10; twoN=2*N;
    int i; for(i=0;i<N;i++)perm[i]=i+1; for(i=0;i<twoN+1;i++)hist[i]=0;
    heap(N);
    long long total=0; int c; for(c=0;c<twoN+1;c++) total+=hist[c];
    printf("total=%lld\n",total);
    return 0;
}
